sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -611,20 +611,33 @@ RT::PiProgram ProgramManager::getBuiltPIProgram(
 
   for (RTDeviceBinaryImage::PropertyRange::ConstIterator It : ARange) {
     using namespace std::literals;
-    if ((*It)->Name != "aspects"sv)
-      continue;
-    ByteArray Aspects = DeviceBinaryProperty(*It).asByteArray();
-    // 8 because we need to skip 64-bits of size of the byte array
-    auto *AIt = reinterpret_cast<const std::uint32_t *>(&Aspects[8]);
-    auto *AEnd =
-        reinterpret_cast<const std::uint32_t *>(&Aspects[0] + Aspects.size());
-    while (AIt != AEnd) {
-      auto Aspect = static_cast<aspect>(*AIt);
-      if (!Dev->has(Aspect))
-        throw sycl::exception(errc::kernel_not_supported,
-                              "Required aspect " + getAspectNameStr(Aspect) +
-                                  " is not supported on the device");
-      ++AIt;
+    if ((*It)->Name == "aspects"sv) {
+      ByteArray Aspects = DeviceBinaryProperty(*It).asByteArray();
+      // 8 because we need to skip 64-bits of size of the byte array
+      Aspects.dropBytes(8);
+      while (!Aspects.empty()) {
+        auto Aspect = static_cast<aspect>(Aspects.consume<int>());
+        if (!Dev->has(Aspect))
+          throw sycl::exception(errc::kernel_not_supported,
+                                "Required aspect " + getAspectNameStr(Aspect) +
+                                    " is not supported on the device");
+      }
+    } else if ((*It)->Name == "reqd_sub_group_size"sv) {
+      ByteArray ReqdSubGroupSize = DeviceBinaryProperty(*It).asByteArray();
+      // Drop 8 bytes describing the size of the byte array.
+      ReqdSubGroupSize.dropBytes(8);
+      auto SupportedSubGroupSizes =
+          Device.get_info<info::device::sub_group_sizes>();
+      while (!ReqdSubGroupSize.empty()) {
```

Comment:
It is expected that there will be only one reqd-sub-group size per device image. Therefore, the loop is a bit confusing here

Moreover, it is probably better to use `uint32` property format instead of `byte_array`, because it would better represent the expected content of the property and would allow to save 8 bytes per device image :)

https://github.com/intel/llvm/blob/a4164ef51365e7122168d1a69cdffc6b5da2d187/llvm/include/llvm/Support/PropertySetIO.h#L61-L62

Reply:
I would vote for (1) - no attribute in source code means no property attached to device image.

(2) is something we _may_ consider in longer term, once we approach implementation of [sycl_ext_oneapi_named_sub_group_sizes](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc). That extension introduces two new special values for `reqd-sub-group-size`, such as `auto` and `primary` and we will need a way to represent them somehow. It could be that we will introduce a new IR/SPIR-V level attributes, or it could be that we will reserve a special values in existing `reqd-sub-group-size` attribute. Either way, unless implementation for named sub-group sizes is agreed upon, I'm hesitant to reserve special values in existing `reqd-sub-group-size` attribute and I think that the lack of it should be represented as the lack of it, at all level in our toolchain; at least in the current implementation.